crates/bevy_ecs/src/query/state.rs

```diff
@@ -1877,6 +1892,76 @@ impl<D: QueryData, F: QueryFilter> From<QueryBuilder<'_, D, F>> for QueryState<D
     }
 }
 
+/// Warpper of [`QueryState`] for safety reasons.
+/// The inner state in manage by observer. DO NOT try to mutate it outside of the observer,
+/// event with unsafe code.
+#[derive(Component)]
+#[component(on_add = on_add_query_state::<D, F>, immutable)]
+pub(crate) struct QueryStateWrapper<D: QueryData, F: QueryFilter>(QueryState<D, F>);
+
+impl<D: QueryData, F: QueryFilter> QueryStateWrapper<D, F> {
+    pub(crate) fn new(query_state: QueryState<D, F>) -> Self {
+        Self(query_state)
+    }
+
+    pub(crate) fn inner(&self) -> &QueryState<D, F> {
+        &self.0
+    }
+}
+
+fn on_add_query_state<D: QueryData + 'static, F: QueryFilter + 'static>(
+    mut world: DeferredWorld,
+    ctx: HookContext,
+) {
+    let entity = ctx.entity;
+
+    let observer = Observer::new(
+        move |trigger: Trigger<ArchetypeCreated>, mut world: DeferredWorld| -> Result {
+            let archetype_id: ArchetypeId = trigger.event().0;
+            let archetype_ptr: *const Archetype = world
+                .archetypes()
+                .get(archetype_id)
+                .expect("Invalid ArchetypeId");
+            let Some(mut state) = world.get_entity_mut(entity).ok().and_then(|entity| unsafe {
+                // SAFETY: This in intended mutation
+                entity.into_mut_assume_mutable::<QueryStateWrapper<D, F>>()
+            }) else {
+                world.commands().entity(trigger.observer()).despawn();
+                return Ok(());
+            };
+
+            // SAFETY: The archetype and state is from the same world
+            unsafe {
+                state.0.new_archetype_internal(&*archetype_ptr);
+            }
+
+            Ok(())
+        },
+    );
+
+    world.commands().queue(move |world: &mut World| {
+        world.spawn(observer);
+        let world = world.as_unsafe_world_cell();
+        let entity_mut = world.get_entity(entity).unwrap();
+        // SAFETY: This is intended mutation
+        let mut state = unsafe {
+            entity_mut
+                .get_mut_assume_mutable::<QueryStateWrapper<D, F>>()
+                .unwrap()
+        };
+        // Its still possible archetype created after the state is created and before getting sync by
+        // the observer. The archetype [Observer, ObserverState] and [QueryState<D, F>] for example.
+        // SAFETY: We keep a mutable reference but this method doesn't read that reference.
+        state.0.update_archetypes_unsafe_world_cell(world);
```

Comment:
I found this nasty bug, it's still possible to miss some archetypes that cause query like `Query<Entity>` miss some data.

Reply:
Oh, that is subtle!  It would be good to add a unit test that fails without this line.